### PR TITLE
Fix KeyError: SecurityGroups in elasticache module. 

### PR DESCRIPTION
**Pull request by stefanhorning**
##### SUMMARY
Also improve docs a little.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
elasticache module

##### ADDITIONAL INFORMATION
Module would otherwise throw `Fix KeyError: 'SecurityGroups'` when updating an existing Cache Cluster with a security group.

Also it was confusing to me at first if the security group param should be provided when in a VPC, as the `cache_security_groups` flag can't be used then and I didn't read further down the docs (on the [Docs website](https://docs.ansible.com/ansible/latest/collections/community/aws/elasticache_module.html#parameter-cache_security_groups)). Hence I added a hint for others. But ideally this should be handled by the module internally, as it can be detected if VPC or not, based on the cache_subnet_group param (so only one SG param needs to be exposed to user).

### CODE REVIEW

File: tests/integration/targets/elasticache/tasks/main.yml
Comment (markuman) [inline tests/integration/targets/elasticache/tasks/main.yml:12]: Can you add an `always` block please, which will always delete elasticache instances in case of errors.

File: tests/integration/targets/elasticache/tasks/main.yml
Comment (markuman) [inline tests/integration/targets/elasticache/tasks/main.yml:15]: use full qualified collection name. `community.aws.elasticache`.

**Review by markuman — CHANGES_REQUESTED** [review]
It would be nice if you can expand the integration test that it tests the SecurityGroup option of your PR.  
Furthermore, you need a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment).

**Comment by stefanhorning** [comment]
Sure, I was in the progress of looking into expanding the tests anzway.

File: tests/integration/targets/elasticache/tasks/main.yml
Comment (stefanhorning) [inline tests/integration/targets/elasticache/tasks/main.yml:17]: Not sure if I need prefixing in my test resource names so they don't conflict with other tests. Also if there is s.th. like a global VPC that can be used by all tests, I am happy to use that. Just wasn't sure were to look these things up.

File: tests/integration/targets/elasticache/tasks/main.yml
Comment (stefanhorning) [inline tests/integration/targets/elasticache/tasks/main.yml:9]: Looks like this whole section isn't working, as params are not applied to my test tasks. What is it I am doing wrong here? I simply took this from other tests.

**Comment by stefanhorning** [comment]
Also I am getting this build error:

"/usr/local/lib/python2.7/dist-packages/OpenSSL/crypto.py:14: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.\n  from cryptography import utils, x509\nTraceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-tmp-1613384043.93-354-274813038104715/AnsiballZ_elasticache.py

Shouldn't we stop testing with Python 2.7, as it's super outdated an mist libraries start dropping support for it.

**Comment by tremble** [comment]
> Shouldn't we stop testing with Python 2.7, as it's super outdated an mist libraries start dropping support for it.

We've mostly picked up our support 'requirements' from Ansible itself.  Which will likely have some form of support for RHEL6 for another couple of years.

Because boto3 is dropping support for Python 2.7 I suspect we'll be releasing a 'major' version somewhere after July in which we drop official support for Python 2.7.  ( https://boto3.amazonaws.com/v1/documentation/api/1.16.56/guide/migrationpy3.html )

**Comment by stefanhorning** [comment]
@tremble So what is the suggested way in the meantime to suppress CI errors like these? Catch the warning in the module, skip python2.7 tests for specific test targets, don't run these tests on CI at all?

**Comment by stefanhorning** [comment]
@markuman Please review again. 

Also if s.b. could address my above comments about the test setup that would also be nice (AWS params, resource prefixes).

**Comment by tremble** [comment]
The python 2.7 **warning** can be ignored.  Shippable stuck everything on one line.  The **error** was that it couldn't find the credentials.

Try using the short name for module instead of the FQCN.  Both **should** work, but I'm guessing a bug somewhere.

File: tests/integration/targets/elasticache/tasks/main.yml
Comment (markuman) [reply tests/integration/targets/elasticache/tasks/main.yml:17]: The worst thing that can happen is, that your integration test will run twice at the same time, that is triggered by another PR later. Think of the other test will delete the vpc resource while the other is still using it.    
Take a look at the integration test of `ec2_instance` for example. It is using: `vpc_name: '{{ resource_prefix }}-vpc'`  
I guess it comes from `meta/` that includes `prepare_tests (_but I'm not 100% sure about it_). Maybe @tremble an clearify it.

Comment (stefanhorning) [reply tests/integration/targets/elasticache/tasks/main.yml:17]: Ok, will just use that, seems like a common pattern at least ;)

Comment (tremble) [reply tests/integration/targets/elasticache/tasks/main.yml]: Yeah, you'll probably hit weird race conditions with the way this is currently laid out.

The way I tend to do it
https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/ec2_vpc_endpoint/tasks/main.yml
https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/ec2_vpc_endpoint/defaults/main.yml
Tries to ensure that neither the name nor the CIDR should be the same between any 2 test runs that might be happening in parallel

**Comment by stefanhorning** [comment]
all code should be fixed now. Now it's just about fixing the CI setup to allow access for test user:

```
User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible-collections=community.aws=1534.24 is not authorized to perform: elasticache:DescribeCacheClusters on resource: arn:aws:elasticache:us-east-1:966509639900:cluster:shippable-1534-24-elasticache-module-redis-test
```

**Comment by markuman** [comment]
@stefanhorning you need another PR in [mattclay/aws-terminator](https://github.com/mattclay/aws-terminator)  
it's a service that deletes all dangle services and resources.  
I guess `elasticache` belongs to the `data-services.yaml`.  
https://github.com/mattclay/aws-terminator/blob/master/aws/policy/data-services.yaml

Just reference in that PR to this PR.

**Comment by stefanhorning** [comment]
I guess I addressed all issues now. So hopefully CI will turn green once https://github.com/mattclay/aws-terminator/pull/126 is merged.

File: tests/integration/targets/elasticache/tasks/main.yml
Comment (jillr) [inline tests/integration/targets/elasticache/tasks/main.yml:80]:
```suggestion
        name: "{{ elasticache_redis_test_name }}"
```
I think this is what's meant here.  Otherwise in `elasticache_redis.name` name is not defined.

File: tests/integration/targets/elasticache/tasks/main.yml
Comment (jillr) [inline tests/integration/targets/elasticache/tasks/main.yml:96]:
```suggestion
          elasticache_redis_new.elasticache.data.SecurityGroups.0.SecurityGroupId ==  elasticache_redis_sg.group_id
```

This one is throwing syntax errors, this passes locally for me

**Review by jillr — CHANGES_REQUESTED** [review]
New policies and terminator will be deployed shortly, couple of things here I found while testing those.

**Comment by stefanhorning** [comment]
@jillr I think I addressed all your comments and a few more issues. I hope test will run through now. Once green this PR can be merged.

**Comment by stefanhorning** [comment]
Tests are running through now. Just get an error in one of the python versions (running in parallel to the others):

```json
"code": "InsufficientCacheClusterCapacity", 
"message": "Cannot create a cache cluster because there is no availability zone corresponding to the subnets in the cache subnet group with sufficient capacity."
```

This is probably due to an AWS limit that get's hit if all the parallel tests try to create a cluster. @jillr Probably we need to increase the cache cluster limits for that AWS account.

File: changelogs/fragments/410-elasticache-fixes.yml
Comment (tremble) [inline changelogs/fragments/410-elasticache-fixes.yml:3]:
```suggestion
  - elasticache - Fix ``KeyError`` issue when updating security group (https://github.com/ansible-collections/community.aws/pull/410).
```

File: changelogs/fragments/410-elasticache-fixes.yml
Comment (tremble) [inline changelogs/fragments/410-elasticache-fixes.yml:5]:
```suggestion
  - elasticache - Improve docs a little, add intgration tests (https://github.com/ansible-collections/community.aws/pull/410).
```

File: tests/integration/targets/elasticache/aliases
Comment (tremble) [inline tests/integration/targets/elasticache/aliases:3]:
```suggestion
# Sometimes hit AWS capacity issues - InsufficientCacheClusterCapacity
# https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/ErrorMessages.html#ErrorMessages.INSUFFICIENT_CACHE_CLUSTER_CAPACITY
unstable

cloud/aws
shippable/aws/group1

elasticache_subnet_group
```

**Review by tremble — APPROVED** [review]
Couple of minor tweaks.

The flake was due to Amazon capacity issues:
https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/ErrorMessages.html#ErrorMessages.INSUFFICIENT_CACHE_CLUSTER_CAPACITY

**Review by jillr — APPROVED** [review]
Tests pass locally and LGTM once CI (excluding shippable) is green

**Comment by tremble** [comment]
recheck

**Pull request merged**
